Linux/fundamental/file_descriptor_non_blocking_IO_and_epoll.md

```diff
@@ -0,0 +1,215 @@
+
+Sự khác biệt giữa 2 chế độ Level-Triggered và Edge-Triggered được thể hiện ở bước (5), khi system call epoll\_wait() thứ 2 được thực hiện. Lúc này, nếu tại thời điểm chúng ta thực hiện epoll\_create() ở bước 1, chế độ notification mà chúng ta truyền vào epoll\_create() là Level-Triggered mode, thì epoll\_wait() ở bước (5) sẽ ngay lập tức trả về kết quả thông báo **rfd** File Descriptor ở trạng thái READY. Lý do là như đã nói ở trên, ở Level-Triggered mode thì bất cứ khi nào I/O resource ở trạng thái READY, kernel sẽ trả về thông báo, tại thời điểm  epoll\_wait() thứ 2 được gọi, chúng ta vẫn còn 2-1 = 1 kB dữ liệu trong read side pipe resource, do vậy I/O resource này vẫn đang ở trạng thái READY. Do đó epoll\_wait() thứ 2 lập tức trả về kết quả là File Descriptor rfd. Còn trong trường hợp tại thời điểm chúng ta thực hiện epoll\_create() ở bước 1, chế độ notification mà chúng ta truyền vào epoll\_create() là Edge-Triggered mode, thì gần như chắc chắn epoll\_wait() sẽ bị block lại. Lý do là như chúng ta đã nói ở trên, trong một chu kỳ thay đổi trạng thái READY, thì ở Edge-Triggered mode kernel chỉ trả về thông báo một lần. Mà ở trong trường hợp này, tại bước (3) kernel đã trả về thông báo một lần rồi, nên tại bước 5, khi chúng ta gọi epoll\_wait() một lần nữa, thì vẫn chưa có sự kiện thay đổi từ NON-READY sang READY thứ hai xảy ra (\*\*), do đó epoll\_wait() ở Edge-Triggered mode sẽ block process này lại cho tới khi xảy ra điều kiện (\*\*). Để điều này xảy ra, chúng ta cần có một process thứ 2 đọc nốt 1 kB dữ liệu trong **read side pipe resource** này để rfd trở về trạng thái NON-READY, sau đó process thứ 2 này thực hiện tiếp việc ghi dữ liệu vào write side of the pipe để đẩy dữ liệu vào **read side pipe resource**. Lúc này do **read side pipe resource** đã có dữ liệu, nên sự kiện event chuyển từ NON-READY sang READY thứ hai xảy ra, và điều này kích hoạt điều kiện (\*\*). Lúc này system call epoll\_wait() ở bước 5 sẽ trả về kết quả cho process thông báo rfd ở trạng thái READY.
+
+Qua phân tích trên, chúng ta có thể thây chúng ta cần cẩn thận khi sử dụng Edge-Triggered mode, tránh tạo ra các dead lock blocking.
+
+Như vậy, chúng ta có thể thấy rằng, Edge-Triggered mode có vẻ phức tạp và khó dùng hơn Level-Triggered Mode. Vậy tại sao chế độ này vẫn được sử dụng trong thực tế, nó có ưu điểm gì? 
+(to be continue)
+
```

Comment:
tốt, merge PR này.
phần (to be continue) tiếp theo bổ sung theo comment của anh ở trên về limit của fd và ví dụ về apache, nginx, mysql.